improver/lapse_rate.py

```diff
@@ -46,76 +45,92 @@
-            lapse_rate.slices([lapse_rate.coord(axis='y'),
-                               lapse_rate.coord(axis='x')]),
-            temperature.slices([temperature.coord(axis='y'),
-                                temperature.coord(axis='x')])):
-
-        # calculate temperature adjustment in K
-        adjustment = multiply(orog_diff, lrsubcube)
-
-        # apply adjustment to each spatial slice of the temperature cube
-        newcube = tempsubcube.copy()
-        newcube.convert_units('K')
-        newcube.data += adjustment.data
-        adjusted_temperature.append(newcube)
-
-    return iris.cube.CubeList(adjusted_temperature).merge_cube()
+class ApplyGriddedLapseRate(PostProcessingPlugin):
+    """Class to apply a lapse rate adjustment to a temperature data forecast"""
+
+    def _check_dim_coords(self, temperature):
+        """Throw an error if the dimension coordinates are not the same for
+        temperature and lapse rate cubes
+
+        Args:
+            temperature (iris.cube.Cube)
+        """
+        for crd in temperature.coords(dim_coords=True):
+            try:
+                if crd != self.lapse_rate.coord(crd.name()):
+                    raise ValueError(
+                        'Lapse rate cube coordinate "{}" does not match '
+                        'temperature cube coordinate'.format(crd.name()))
+            except CoordinateNotFoundError:
+                raise ValueError('Lapse rate cube has no coordinate '
+                                 '"{}"'.format(crd.name()))
+
+    def _calc_orog_diff(self, source_orog, dest_orog):
+        """Get difference in orography heights, in metres
+
+        Args:
+            source_orog (iris.cube.Cube):
+                2D cube of source orography heights (units modified in place)
+            dest_orog (iris.cube.Cube):
+                2D cube of destination orography heights (units modified in
+                place)
+
+        Returns:
+            iris.cube.Cube
+        """
+        source_orog.convert_units('m')
+        dest_orog.convert_units('m')
+        orog_diff = (next(dest_orog.slices(self.xy_coords)) -
+                     next(source_orog.slices(self.xy_coords)))
+        return orog_diff
+
+    def process(self, temperature, lapse_rate, source_orog, dest_orog):
+        """Applies lapse rate correction to temperature forecast.  All cubes'
+        units are modified in place.
+
+        Args:
+            temperature (iris.cube.Cube):
+                Input temperature field to be adjusted
+            lapse_rate (iris.cube.Cube):
+                Cube of pre-calculated lapse rates
+            source_orog (iris.cube.Cube):
+                2D cube of source orography heights
+            dest_orog (iris.cube.Cube):
+                2D cube of destination orography heights
+
+        Returns:
+            iris.cube.Cube:
+                Lapse-rate adjusted temperature field
+        """
+        self.lapse_rate = lapse_rate
```

Comment:
If this is being set to self like this, why isn't this passed into an init?

Reply:
Because Ben told me not to (see above).
This makes sense to set up as a class member, as it's conceptually relevant to the class "meaning" (to apply a lapse rate), and is used in a lot of places inside the class so isn't really worth passing around.
I'd initially designed it your way with an init, but Ben wanted it elsewise and I didn't really feel like it was worth arguing over.  Happy to change it back though ... ;)